### PR TITLE
EES-2363 Add summary field to ReleaseFile and allow updating via endpoint

### PR DESCRIPTION
**Pull request by ntsim**
This PR introduces a new `Summary` field to `ReleaseFile` entities, which can then be updated via a new update endpoint (using `PATCH /release/{releaseId}/files/{fileId}`). This update endpoint supersedes the previous endpoint which only updated file names.

We have implemented this as a PATCH as it's not necessary for all files to have to update their name/title or summary. These fields are optional in the request, and will not unset the relevant entity's properties if set to null. 

If you wish to set these fields to empty values - use empty strings.

## Related changes

- Adds missing various checks to `ReleaseFileService.GetFile` to ensure that the user is allowed to view the release and the associated file.
- Consolidates `AncillaryFileInfo` into `FileInfo` as it only contains common properties that are not unique to ancillary files.
- Various cleanup across `ReleaseFileRepository` get/update methods.

## Other changes

- Renamed various `Name` fields to `Title` in upload/update requests. This is in preparation for further alignment of this field name across the frontend and backend (see [EES-2517](https://dfedigital.atlassian.net/browse/EES-2517)) for consistency.

### CODE REVIEW

File: src/GovUk.Education.ExploreEducationStatistics.Admin.Tests/Services/ReleaseFileServicePermissionTests.cs
Comment (duncan-at-hiveit) [inline src/GovUk.Education.ExploreEducationStatistics.Admin.Tests/Services/ReleaseFileServicePermissionTests.cs:174]: Didn't know of this short-hand before.  Worth mentioning in #development channel

File: src/GovUk.Education.ExploreEducationStatistics.Admin.Tests/Services/ReleaseFileServiceTests.cs
Comment (duncan-at-hiveit) [inline src/GovUk.Education.ExploreEducationStatistics.Admin.Tests/Services/ReleaseFileServiceTests.cs:1493]: Can now use `result.AssertNotFound()` which'll do these 2 checks in one.

File: src/GovUk.Education.ExploreEducationStatistics.Admin.Tests/Services/ReleaseFileServiceTests.cs
Comment (duncan-at-hiveit) [inline src/GovUk.Education.ExploreEducationStatistics.Admin.Tests/Services/ReleaseFileServiceTests.cs:1827]: Doesn't really gain you anything here, but just so you're aware, you can also do:

`var unit = result.AssertRight()`

To both assert and extract the contents of the right in one go.

File: src/GovUk.Education.ExploreEducationStatistics.Admin.Tests/Services/ReleaseFileServiceTests.cs
Comment (duncan-at-hiveit) [inline src/GovUk.Education.ExploreEducationStatistics.Admin.Tests/Services/ReleaseFileServiceTests.cs:2065]: Now this'd be a better candidate for `var upload = result.AssertRight();`

File: src/GovUk.Education.ExploreEducationStatistics.Admin/Controllers/Api/ReleasesController.cs
Comment (duncan-at-hiveit) [inline src/GovUk.Education.ExploreEducationStatistics.Admin/Controllers/Api/ReleasesController.cs:104]: Is there a reason we're doing this manually like this and not being handed a `AddDataFileViewModel` or similar, as we do in other places?  Like we currently use `ReleaseAncillaryFileUploadViewModel`?  I'm aware this is pre-existing code and not something you've added - just wondering if there was a reason...

File: src/GovUk.Education.ExploreEducationStatistics.Admin/Models/DataFileInfo.cs
Comment (duncan-at-hiveit) [inline src/GovUk.Education.ExploreEducationStatistics.Admin/Models/DataFileInfo.cs:13]: Why the need for the `new` keyword here, I wonder?  Maybe you don't need it with `const`?

File: src/GovUk.Education.ExploreEducationStatistics.Admin/Services/ReleaseFileRepository.cs
Comment (duncan-at-hiveit) [inline src/GovUk.Education.ExploreEducationStatistics.Admin/Services/ReleaseFileRepository.cs:105]: I think we already have this Find / FindOrNotFound pattern for free within AbstractRepository<T> - you could look to extend it and see if it gives you what you need.  It might give you more than you need too!

File: src/GovUk.Education.ExploreEducationStatistics.Content.Model/ReleaseFile.cs
Comment (duncan-at-hiveit) [inline src/GovUk.Education.ExploreEducationStatistics.Content.Model/ReleaseFile.cs:12]: Why the need for the exclamation mark in this syntax?  If it's possible to set this to null, then should it be `public Release?` rather than `public Release`   ?

Then presumably you wouldn't have to tell the compiler to be forgiving?

File: src/GovUk.Education.ExploreEducationStatistics.Admin/Controllers/Api/ReleasesController.cs
Comment (duncan-at-hiveit) [reply src/GovUk.Education.ExploreEducationStatistics.Admin/Controllers/Api/ReleasesController.cs:104]: Oh, I've answered my own question - query params.  Which changes my question to, is there any preference in using query params in POSTs over a request body?  Understandably we need the body wherever we upload Files, but just personal preference for me is to use a body all over the place.

Anyway, nowt to do with this PR, so forget it :)

Comment (ntsim) [reply src/GovUk.Education.ExploreEducationStatistics.Admin/Controllers/Api/ReleasesController.cs:104]: Yep, so I think this is just technical debt from when this was originally written. We really should be binding it to a model using the `[FromForm]` annotation like we've done in the ancillary file upload for this PR). 

I decided against doing this as part of this ticket as there's a bunch of things to clean up here :smile:

I've created this ticket for it: https://dfedigital.atlassian.net/browse/EES-2521

**Review by duncan-at-hiveit — DISMISSED** [review]
Left a few comments and questions, but nothing that should hold this up until I'm back in on Friday, so approving now, and feel free to check out the comments.  Ta.

File: src/GovUk.Education.ExploreEducationStatistics.Admin/Services/ReleaseFileRepository.cs
Comment (ntsim) [reply src/GovUk.Education.ExploreEducationStatistics.Admin/Services/ReleaseFileRepository.cs:105]: I thought about this, but unfortunately it's only in the `Data.Services` project and is currently specific for the `StatisticsDbContext`. Thought it best to avoid re-purposing it here as it's used in a number of places.

It would be good if we could create a common base repository class for stuff like this that we could have in `Common`. I've created a ticket for it anyway: https://dfedigital.atlassian.net/browse/EES-2529

File: src/GovUk.Education.ExploreEducationStatistics.Content.Model/ReleaseFile.cs
Comment (ntsim) [reply src/GovUk.Education.ExploreEducationStatistics.Content.Model/ReleaseFile.cs:12]: As we have enabled nullables, reference types need to be initialised upon construction, however, this really requires you to have explicit constructor methods that initialize everything correctly.  Unfortunately, this doesn't play well with Entity Framework as I _think_ it typically uses setters to construct everything, hence we need an empty constructor (hence fields are potentially null upon construction).

I believe a similar thing happens with Hibernate, so you typically need to have an empty constructor method.

So basically, a workaround for this is to use the null-forgiving operator to basically re-assure the C# compiler that everything is fine. By using it we say that our field will be initialized before we actually use it - by something like Entity Framework.

File: src/GovUk.Education.ExploreEducationStatistics.Admin/Models/DataFileInfo.cs
Comment (ntsim) [reply src/GovUk.Education.ExploreEducationStatistics.Admin/Models/DataFileInfo.cs:13]: Without `new`, we get this compiler warning:

> The keyword 'new' is required on 'Type' because it hides property 'FileType GovUk.Education.ExploreEducationStatistics.Common.Model.FileInfo.Type'

Given it's the same signature as the base class, you'd think it would be fine as it's not really hiding anything, but hey, I don't make the compiler rules :man_shrugging: 

I didn't really need to change it to `const` either, but I originally got the suggestion in another case that was similar to this. Seemed to make sense so thought I'd throw it in here too!

File: src/GovUk.Education.ExploreEducationStatistics.Admin.Tests/Services/ReleaseFileServiceTests.cs
Comment (ntsim) [reply src/GovUk.Education.ExploreEducationStatistics.Admin.Tests/Services/ReleaseFileServiceTests.cs:1493]: Done!

File: src/GovUk.Education.ExploreEducationStatistics.Admin.Tests/Services/ReleaseFileServiceTests.cs
Comment (ntsim) [reply src/GovUk.Education.ExploreEducationStatistics.Admin.Tests/Services/ReleaseFileServiceTests.cs:1827]: Done!

File: src/GovUk.Education.ExploreEducationStatistics.Content.Model/ReleaseFile.cs
Comment (duncan-at-hiveit) [reply src/GovUk.Education.ExploreEducationStatistics.Content.Model/ReleaseFile.cs:12]: Good stuff, thanks!

File: src/GovUk.Education.ExploreEducationStatistics.Admin/Services/ReleaseFileRepository.cs
Comment (duncan-at-hiveit) [reply src/GovUk.Education.ExploreEducationStatistics.Admin/Services/ReleaseFileRepository.cs:105]: Cool ta

File: src/GovUk.Education.ExploreEducationStatistics.Admin/Models/DataFileInfo.cs
Comment (duncan-at-hiveit) [reply src/GovUk.Education.ExploreEducationStatistics.Admin/Models/DataFileInfo.cs:13]: Yeaqh I would've thought it would have been fine too, but meh!  Cool thanks.

File: src/GovUk.Education.ExploreEducationStatistics.Admin/Controllers/Api/ReleasesController.cs
Comment (duncan-at-hiveit) [reply src/GovUk.Education.ExploreEducationStatistics.Admin/Controllers/Api/ReleasesController.cs:104]: Yup understandable, I knew it was out of the remit of this PR, but thanks for creating the ticket!

**Review by duncan-at-hiveit — APPROVED** [review]

**Pull request merged**
